src/dataframe/dataframe.jl

```diff
@@ -278,6 +278,26 @@ _columns(df::DataFrame) = getfield(df, :columns)
 nrow(df::DataFrame) = ncol(df) > 0 ? length(_columns(df)[1])::Int : 0
 ncol(df::DataFrame) = length(index(df))
 
+##############################################################################
+##
+## DataFrame consistency check
+##
+##############################################################################
+
+function _check_consistency(df::DataFrame)
+    cols, idx = _columns(df), index(df)
+    ncols = length(cols)
+    @assert length(idx.names) == length(idx.lookup) == ncols "Index and columns of a data frame do not match"
+    ncols == 0 && return nothing
+    nrows = length(cols[1])
+    for i in 2:length(cols)
+        @assert length(cols[i]) == nrows "Length of column $i does not match length of column 1"
+    end
+    nothing
+end
+
+_check_consistency(df::AbstractDataFrame) = _check_consistency(parent(df))
```

Comment:
This doesn't sound fully generic, as we don't require `AbstractDataFrame` to have a parent (even if that's the case in practice). Is there a way to check consistency only using `AbstractDataFrame` API?

Reply:
Actually I think that `parent` is part of `AbstractDataFrame` API (like it is part of arrays API in Base).
But if you think it is not then I can just special case for `SubDataFrame`. This is not a big deal whatever we choose I think 😄.